### ISSUE TITLE
Benchmark tipset validation bottlenecks

### ISSUE THREAD

**Issue by lemmih** [ChainSafe/forest#2048]
**Issue summary**

Forest was recently refactored to remove a network bottleneck when syncing and CPU performance is now the limiting factor. When syncing on my machine, only around three cores are in active use. Ideally we would like to use all available cores.

Tipsets are validated in sequence starting from lower epochs going towards higher epochs. Each tipset contains a set of blocks which are validated concurrently. Each block contains a few checks that are run in sequence and a collection of checks that are run in parallel. See `validate_tipset` and `validate_block` for details.

Tasks:
 - Measure the sequential parts of `validate_block` (ie everything that isn't computed in a separate task).
 - Measure each of the checks that are computed in separate tasks. There are six of them.
 - Ensure the blocks are actually validated concurrently. It is possible that a lock could force them to be validated in sequence.
 - Create a new ticket with an overview of how much time each validation step takes.

Many validations do not require the tipsets to be validated in order. For example, checking that signatures are valid can be done in any order. Other validations write data to the database and do require strict ordering of the tipsets. It might be possible to improve performance by separating the validations that require strict ordering and the ones that do not. However, it could be that the non-ordered validations take up an insignificant amount of time.

The validation step for computing the state root and receipt root cannot be done out-of-order and it is likely to be the most expensive step. We need solid data to tell either way.

**Other information and links**

https://github.com/ChainSafe/forest/blob/5771cc68412742ff06faa5333f9a585d5060de70/blockchain/chain_sync/src/tipset_syncer.rs#L1125
https://github.com/ChainSafe/forest/blob/5771cc68412742ff06faa5333f9a585d5060de70/blockchain/chain_sync/src/tipset_syncer.rs#L1202

**Comment by hanabi1224** [comment]
Just did some analysis by adding `tracing::span` to those functions, then running `cargo run --release -- --chain calibnet --encrypt-keystore false --import-snapshot ~/.local/share/forest/snapshots/calibnet/forest_snapshot_calibnet_2022-10-19_height_1402626.car`

`tipset_sync` takes `31.04s` in total on my machine,
`validate_tipset` takes `6.65s` in total, `0.0087s` on average, while
`validate_block` takes `20.45s` in total, `0.13s` on average. (total is greater than `validate_tipset` because of parralellization)

**Comment by lemmih** [comment]
Testing on mainnet is likely to give very different results than calibnet, btw.

**Comment by lemmih** [comment]
Note that the primary purpose of this issue is to find opportunities for increased parallelism. There's another issue looking at the sequential performance of tipset validation: #2128.

**Comment by lemmih** [comment]
@hanabi1224 Let me know if you wanna talk about how to move this issue forward.

**Comment by hanabi1224** [comment]
@lemmih My major blocker was failure to download mainnet snapshot, I have managed to download one during the weekend with `aria2`, will continue the work this week and let u know if I hit bumps
